pkg/util/active_groups.go

```diff
@@ -0,0 +1,142 @@
+	}
+}
+
+func (ag *ActiveGroups) UpdateGroupTimestampForUser(userID, group string, ts int64) {
+	// Create new atomic before lock is acquired
+	newAtomic := atomic.NewInt64(ts)
+
+	ag.mu.RLock()
+	groupTimestamps := ag.timestampsPerUser[userID]
+	ag.mu.RUnlock()
+
+	if groupTimestamps == nil {
+		ag.mu.Lock()
+		ag.timestampsPerUser[userID] = make(map[string]*atomic.Int64)
+		ag.timestampsPerUser[userID][group] = newAtomic
+		ag.mu.Unlock()
+		return
+	}
+
+	ag.mu.Lock()
+	groupTs := ag.timestampsPerUser[userID][group]
+	if groupTs != nil {
+		ag.mu.Unlock()
+
+		groupTs.Store(ts)
+		return
+	}
+
+	ag.timestampsPerUser[userID][group] = newAtomic
+	ag.mu.Unlock()
+}
+
+func (ag *ActiveGroups) PurgeInactiveGroupsForUser(userID string, deadline int64) []string {
+	ag.mu.RLock()
+	inactiveGroups := make([]string, 0, len(ag.timestampsPerUser))
+	groupTimestamps := ag.timestampsPerUser[userID]
+
+	for group, ts := range groupTimestamps {
+		if ts.Load() <= deadline {
+			inactiveGroups = append(inactiveGroups, group)
+		}
+	}
+	ag.mu.RUnlock()
+
+	if groupTimestamps == nil || len(inactiveGroups) == 0 {
+		return nil
+	}
+
+	// Cleanup inactive groups
+	for i := 0; i < len(inactiveGroups); {
+		group := inactiveGroups[i]
+		deleted := false
+
+		ag.mu.Lock()
+		groupTs := ag.timestampsPerUser[userID][group]
+		if groupTs != nil && groupTs.Load() <= deadline {
+			delete(ag.timestampsPerUser[userID], group)
+			deleted = true
+		}
+		ag.mu.Unlock()
+
+		if deleted {
+			i++
+		} else {
+			inactiveGroups = append(inactiveGroups[:i], inactiveGroups[i+1:]...)
+		}
+	}
+
+	return inactiveGroups
+}
+
+type ActiveGroupsCleanupService struct {
+	services.Service
+
+	activeGroups    *ActiveGroups
+	cleanupFunc     func(string, string) // Takes userID, group
+	inactiveTimeout time.Duration
+}
+
+func NewActiveGroupsCleanupService(cleanupInterval, inactiveTimeout time.Duration, cleanupFn func(string, string)) *ActiveGroupsCleanupService {
+	s := &ActiveGroupsCleanupService{
+		activeGroups:    NewActiveGroups(),
+		cleanupFunc:     cleanupFn,
+		inactiveTimeout: inactiveTimeout,
+	}
+
+	s.Service = services.NewTimerService(cleanupInterval, nil, s.iteration, nil).WithName("active groups cleanup")
+	return s
+}
+
+func NewActiveGroupsCleanupWithDefaultValues(cleanupFn func(string, string)) *ActiveGroupsCleanupService {
+	return NewActiveGroupsCleanupService(1*time.Minute, 20*time.Minute, cleanupFn)
+}
+
+func (s *ActiveGroupsCleanupService) UpdateGroupTimestamp(user, group string, now time.Time) {
+	s.activeGroups.UpdateGroupTimestampForUser(user, group, now.UnixNano())
+}
+
+func (s *ActiveGroupsCleanupService) ActiveGroupLimitExceeded(userID, group string) bool {
+	if s.activeGroups == nil {
+		return false
+	}
+
+	s.activeGroups.mu.RLock()
+	defer s.activeGroups.mu.RUnlock()
+
+	_, containsGroup := s.activeGroups.timestampsPerUser[userID][group]
+	return !containsGroup && len(s.activeGroups.timestampsPerUser[userID]) >= maxGroupsPerUser
+}
```

Comment:
I don't think this method/logic belongs here: why the cleanup service is peeking into `ActiveGroups` tracker internals?